File: sdk/swift/Vil.swift
// =============================================================================
// VIL SDK for Swift — Single-file, zero-dependency SDK.
//
// Usage:
//   swift main.swift                                  — compiles via `vil compile`
//   VIL_COMPILE_MODE=manifest swift main.swift        — prints YAML manifest
//
// API:
//   let p = VilPipeline("name", port: 3080)
//       .sink("webhook", port: 3080, path: "/trigger")
//       .source("inference", url: "http://...", format: "sse", jsonTap: "choices[0].delta.content")
//       .route("webhook.trigger_out", to: "inference.trigger_in", mode: "LoanWrite")
//   p.compile()
// =============================================================================

import Foundation

class ServiceProcess {
    let name: String
    let prefix: String
    struct Ep { let method: String; let path: String; let handler: String }
    var endpoints: [Ep] = []

    init(_ name: String, prefix: String? = nil) {
        self.name = name
        self.prefix = prefix ?? "/api/\(name)"
    }

    @discardableResult
    func endpoint(_ method: String, _ path: String, handler: String? = nil) -> ServiceProcess {
        let h = handler ?? "\(method.lowercased())_\(path.trimmingCharacters(in: CharacterSet(charactersIn: "/")).replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: ":", with: ""))"
        endpoints.append(Ep(method: method, path: path, handler: h))
        return self
    }
}

class VilPipeline {
    private let name: String
    private var port: Int
    private struct Node { let type_: String; let port: Int; let path: String?; let url: String?; let format: String?; let jsonTap: String?; let dialect: String? }
    private struct Route { let from: String; let to: String; let mode: String }
    private var nodeOrder: [String] = []
    private var nodes: [String: Node] = [:]
    private var routes: [Route] = []

    init(_ name: String, port: Int) { self.name = name; self.port = port }

    @discardableResult
    func sink(_ nodeName: String, port: Int = 3080, path: String = "/trigger") -> VilPipeline {
        nodes[nodeName] = Node(type_: "http_sink", port: port, path: path, url: nil, format: nil, jsonTap: nil, dialect: nil)
        if !nodeOrder.contains(nodeName) { nodeOrder.append(nodeName) }
        self.port = port
        return self
    }

    @discardableResult
    func source(_ nodeName: String, url: String, format: String = "sse",
                jsonTap: String? = nil, dialect: String? = nil) -> VilPipeline {
        nodes[nodeName] = Node(type_: "http_source", port: 0, path: nil, url: url, format: format, jsonTap: jsonTap, dialect: dialect)
        if !nodeOrder.contains(nodeName) { nodeOrder.append(nodeName) }
        return self
    }

    @discardableResult
    func route(_ from: String, to: String, mode: String) -> VilPipeline {
        routes.append(Route(from: from, to: to, mode: mode))
        return self
    }

    func toYaml() -> String {
        var lines: [String] = []
        lines.append("vil_version: \"6.0.0\"")
        lines.append("name: \(name)")
        lines.append("port: \(port)")
        lines.append("token: shm")
        if !nodeOrder.isEmpty {
            lines.append(""); lines.append("nodes:")
            for n in nodeOrder {
                let nd = nodes[n]!
                lines.append("  \(n):")
                lines.append("    type: \(nd.type_)")
                if nd.port != 0 { lines.append("    port: \(nd.port)") }
                if let p = nd.path { lines.append("    path: \"\(p)\"") }
                if let u = nd.url { lines.append("    url: \"\(u)\"") }
                if let f = nd.format { lines.append("    format: \(f)") }
                if let j = nd.jsonTap { lines.append("    json_tap: \"\(j)\"") }
                if let d = nd.dialect { lines.append("    dialect: \(d)") }
            }
        }
        if !routes.isEmpty {
            lines.append(""); lines.append("routes:")
            for r in routes {
                lines.append("  - from: \(r.from)")
                lines.append("    to: \(r.to)")
                lines.append("    mode: \(r.mode)")
            }
        }
        return lines.joined(separator: "\n") + "\n"
    }

    func compile() {
        let yaml = toYaml()
        if ProcessInfo.processInfo.environment["VIL_COMPILE_MODE"] == "manifest" {
            print(yaml, terminator: ""); return
        }
        print("  Compiling: \(name)")
        print(yaml, terminator: "")
    }
}

class VilServer {
    private let name: String
    private let port: Int
    private var services: [ServiceProcess] = []
    private struct SemType { let name: String; let kind: String }
    private var semanticTypes: [SemType] = []

    init(_ name: String, port: Int) { self.name = name; self.port = port }

    @discardableResult
    func service(_ svc: ServiceProcess) -> VilServer { services.append(svc); return self }

    @discardableResult
    func semanticType(_ name: String, kind: String) -> VilServer { semanticTypes.append(SemType(name: name, kind: kind)); return self }

    func toYaml() -> String {
        var lines: [String] = []
        lines.append("vil_version: \"6.0.0\"")
        lines.append("name: \(name)")
        lines.append("port: \(port)")
        if !semanticTypes.isEmpty {
            lines.append("semantic_types:")
            for st in semanticTypes { lines.append("  - name: \(st.name)"); lines.append("    kind: \(st.kind)") }
        }
        if !services.isEmpty {
            lines.append("services:")
            for svc in services {
                lines.append("  - name: \(svc.name)")
                lines.append("    prefix: \"\(svc.prefix)\"")
                if !svc.endpoints.isEmpty {
                    lines.append("    endpoints:")
                    for ep in svc.endpoints {
                        lines.append("      - method: \(ep.method)")
                        lines.append("        path: \"\(ep.path)\"")
                        lines.append("        handler: \(ep.handler)")
                    }
                }
            }
        }
        return lines.joined(separator: "\n") + "\n"
    }

    func compile() {
        let yaml = toYaml()
        if ProcessInfo.processInfo.environment["VIL_COMPILE_MODE"] == "manifest" {
            print(yaml, terminator: ""); return
        }
        print("  Compiling: \(name)")
        print(yaml, terminator: "")
    }
}

File: sdk/zig/vil.zig
// =============================================================================
// VIL SDK for Zig — Single-file, zero-dependency SDK.
//
// Usage:
//   zig run main.zig                                  — compiles via `vil compile`
//   VIL_COMPILE_MODE=manifest zig run main.zig        — prints YAML manifest
//
// API:
//   var p = VilPipeline.init("name", 3080);
//   p.sink("webhook", 3080, "/trigger");
//   p.source("inference", "http://...", .{ .format = "sse", .json_tap = "..." });
//   p.route("webhook.trigger_out", "inference.trigger_in", "LoanWrite");
//   p.compile();
// =============================================================================

const std = @import("std");

const MAX_NODES = 16;
const MAX_ROUTES = 32;

pub const SourceOpts = struct {
    format: ?[]const u8 = "sse",
    json_tap: ?[]const u8 = null,
    dialect: ?[]const u8 = null,
};

pub const Node = struct {
    name: []const u8,
    type_: []const u8,
    port: u16,
    path: ?[]const u8,
    url: ?[]const u8,
    format: ?[]const u8,
    json_tap: ?[]const u8,
    dialect: ?[]const u8,
};

pub const Route = struct {
    from: []const u8,
    to: []const u8,
    mode: []const u8,
};

pub const VilPipeline = struct {
    name: []const u8,
    port: u16,
    nodes: [MAX_NODES]Node = undefined,
    node_count: usize = 0,
    routes: [MAX_ROUTES]Route = undefined,
    route_count: usize = 0,

    pub fn init(name: []const u8, port: u16) VilPipeline {
        return VilPipeline{ .name = name, .port = port };
    }

    pub fn sink(self: *VilPipeline, node_name: []const u8, port: u16, path: []const u8) void {
        if (self.node_count < MAX_NODES) {
            self.nodes[self.node_count] = Node{
                .name = node_name, .type_ = "http_sink", .port = port, .path = path,
                .url = null, .format = null, .json_tap = null, .dialect = null,
            };
            self.node_count += 1;
            self.port = port;
        }
    }

    pub fn source(self: *VilPipeline, node_name: []const u8, url: []const u8, opts: SourceOpts) void {
        if (self.node_count < MAX_NODES) {
            self.nodes[self.node_count] = Node{
                .name = node_name, .type_ = "http_source", .port = 0, .path = null,
                .url = url, .format = opts.format, .json_tap = opts.json_tap, .dialect = opts.dialect,
            };
            self.node_count += 1;
        }
    }

    pub fn route(self: *VilPipeline, from: []const u8, to: []const u8, mode: []const u8) void {
        if (self.route_count < MAX_ROUTES) {
            self.routes[self.route_count] = Route{ .from = from, .to = to, .mode = mode };
            self.route_count += 1;
        }
    }

    pub fn compile(self: *VilPipeline) void {
        const stdout = std.io.getStdOut().writer();
        const env = std.posix.getenv("VIL_COMPILE_MODE");
        const is_manifest = if (env) |v| std.mem.eql(u8, v, "manifest") else false;

        self.writeYaml(stdout) catch {};
        if (!is_manifest) {
            stdout.print("  Compiling: {s}\n", .{self.name}) catch {};
        }
    }

    pub fn writeYaml(self: *VilPipeline, writer: anytype) !void {
        try writer.print("vil_version: \"6.0.0\"\nname: {s}\nport: {d}\ntoken: shm\n", .{ self.name, self.port });
        if (self.node_count > 0) {
            try writer.writeAll("\nnodes:\n");
            for (0..self.node_count) |i| {
                const nd = self.nodes[i];
                try writer.print("  {s}:\n    type: {s}\n", .{ nd.name, nd.type_ });
                if (nd.port != 0) try writer.print("    port: {d}\n", .{nd.port});
                if (nd.path) |p| try writer.print("    path: \"{s}\"\n", .{p});
                if (nd.url) |u| try writer.print("    url: \"{s}\"\n", .{u});
                if (nd.format) |f| try writer.print("    format: {s}\n", .{f});
                if (nd.json_tap) |j| try writer.print("    json_tap: \"{s}\"\n", .{j});
                if (nd.dialect) |d| try writer.print("    dialect: {s}\n", .{d});
            }
        }
        if (self.route_count > 0) {
            try writer.writeAll("\nroutes:\n");
            for (0..self.route_count) |i| {
                const r = self.routes[i];
                try writer.print("  - from: {s}\n    to: {s}\n    mode: {s}\n", .{ r.from, r.to, r.mode });
            }
        }
    }
};

pub const ServiceEndpoint = struct {
    method: []const u8,
    path: []const u8,
    handler: []const u8,
};

pub const ServiceProcess = struct {
    name: []const u8,
    prefix: []const u8,
    endpoints: [MAX_NODES]ServiceEndpoint = undefined,
    endpoint_count: usize = 0,

    pub fn init(name: []const u8, prefix: []const u8) ServiceProcess {
        return ServiceProcess{ .name = name, .prefix = prefix };
    }

    pub fn endpoint(self: *ServiceProcess, method: []const u8, path: []const u8, handler: []const u8) void {
        if (self.endpoint_count < MAX_NODES) {
            self.endpoints[self.endpoint_count] = ServiceEndpoint{ .method = method, .path = path, .handler = handler };
            self.endpoint_count += 1;
        }
    }
};

pub const VilServer = struct {
    name: []const u8,
    port: u16,

    pub fn init(name: []const u8, port: u16) VilServer {
        return VilServer{ .name = name, .port = port };
    }

    pub fn compile(self: *VilServer) void {
        const stdout = std.io.getStdOut().writer();
        const env = std.posix.getenv("VIL_COMPILE_MODE");
        const is_manifest = if (env) |v| std.mem.eql(u8, v, "manifest") else false;
        stdout.print("vil_version: \"6.0.0\"\nname: {s}\nport: {d}\n", .{ self.name, self.port }) catch {};
        if (!is_manifest) {
            stdout.print("  Compiling: {s}\n", .{self.name}) catch {};
        }
    }
};
